rfd/0065-distributed-tracing.md

```diff
@@ -0,0 +1,247 @@
+
+#### Manual Instrumentation
+
+In order to instrument packages which don't already have a library available (e.g. golang.org/x/crypto/ssh) we will
+have to manually create them. All Teleport application code will also have to be manually instrumented. In addition to
+adding instrumentation to service boundaries, we will first trace the code paths along `tsh ssh` and `tsh ls` as these
+are two areas which have had the most scrutiny about latency by customers. From there further instrumentation can be
+added to other areas of the codebase, using the code already instrumented as a guide for best practices.
+
+All spans created to instrument Teleport shall follow the guidelines laid out in the
+[spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span).
+Likewise, all attributes shall follow the [guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attribute)
+to provide consistency within Teleport and across other services.
+
+#### Exporting Spans
+
+All spans created by Teleport will be exported in the
+[OpenTelemetry Protocol (OTLP)](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md)
+format. OTLP is a vendor neutral binary format which is used to export traces and metrics to a variety of
+different backends. Due to the fact that different Teleport customers will have different telemetry backends this
+provides us a way to export traces in such a way that they can be easily consumed all customers. Most of the major
+telemetry backends have already, or are working on adding support to consume OTLP spans. Customers also have the option
+to use the [OpenTelemetry Collector (otel-collector)](https://github.com/open-telemetry/opentelemetry-collector) to
+receive, process, and export the traces from Teleport to their desired backend. This is particularly useful for telemetry
+backends which don't currently support OTLP. The otel-collector could be configured to retrieve spans from Teleport and
+then forward them the backend in a format that the backend supports.
+
+
+##### Tracing Configuration
+```yaml
+tracing:
+  # the url of the exporter to send spans to - can be http/https/grpc
+  exporter_url: "http://localhost:14268/api/traces"
+  # the number of spans to sample per 100
+  sampling_rate_per_hundred: 10
+```
+
+Teleport would consume the above configuration to create a [`TracerProvider`](https://pkg.go.dev/go.opentelemetry.io/otel/trace#TracerProvider)
+and [`SpanExporter`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#SpanExporter) which would only record 10% of
+all spans and export them to the specified url.
+
+
+#### Exporting Spans from tsh/tctl/tbot
+
+In order to propagate spans that originated from `tctl`, `tsh`, or `tbot` we have two options:
+
+1) Export the spans directly into the telemetry backend
+2) Make the Auth Server a span forwarder
```

Comment:
I strongly agree with pursuing the second option here. Exporting spans directly to the telemetry backend raises a lot of issues around preventing abuse and authentication, especially if we want to use this within Teleport cloud.